mesop/components/markdown/markdown.ts

```diff
@@ -17,14 +27,79 @@ export class MarkdownComponent {
   @Input() key!: Key;
   @Input() style!: Style;
 
+  htmlOutput: SafeHtml = '';
+
+  private _codeBlocks: string[] = [];
   private _config!: MarkdownType;
 
-  constructor(readonly changeDetectorRef: ChangeDetectorRef) {}
+  constructor(
+    readonly changeDetectorRef: ChangeDetectorRef,
+    private sanitizer: DomSanitizer,
+    private renderer: Renderer2,
+    private el: ElementRef,
+  ) {}
 
   async ngOnChanges() {
     this._config = MarkdownType.deserializeBinary(
       this.type.getValue() as unknown as Uint8Array,
     );
+
+    // Create a custom renderer so we can add highlighting and a copy button to code
+    // blocks.
+    //
+    // We recreate the renderer each time the markdown gets updated, since we need to
+    // keep track of the code blocks for the copy button.
+    //
+    const renderer = new marked.Renderer();
+    this._codeBlocks = [];
+    // We do not put the raw code in the HTML markup. Instead we just add the index of
+    // the raw code to look up on click.
+    let index = -1;
+    renderer.code = ({text, lang}) => {
+      const language =
+        lang !== undefined && hljs.getLanguage(lang) ? lang : 'plaintext';
+      const highlighted = hljs.highlight(text, {language}).value;
+      this._codeBlocks.push(text);
+      index += 1;
+      // We use a <div class="code-${index}"></div> as a placeholder since Angular does
+      // not allow buttons during HTML sanitization.
+      return `<div class="code-block"><div class="code-${index}"></div><pre><code class="hljs ${language}">${highlighted}</code></pre></div>`;
+    };
+
+    // Use the custom renderer in Marked.js
+    marked.use({renderer});
+
+    const output = await marked(this._config!.getText()!, {gfm: true});
+
+    // Sanitize the generated HTML from MarkedJS first.
+    const html = this.sanitizer.sanitize(SecurityContext.HTML, output) || '';
+    // Now update the placeholder div with our button once we know the output is
+    // sanitized.
+    this.htmlOutput = this.sanitizer.bypassSecurityTrustHtml(
```

Comment:
you might not be able to call this method downstream b/c of security checks and I don't think you need it. You should be able to directly replace the HTML from L72 with the `<button>` and then have this.htmlOutput be a string. When you bind htmlOutput to innerHTML, Angular should implicitly apply the sanitizer and it should leave your button code alone since it doesn't actually contain JS.

Reply:
Unfortunately, Angular will sanitize the button, but I just changed it to an "a" tag and that worked.